### PR TITLE
git: add tracer package

### PR DESCRIPTION
**Pull request by aymanbagabas**
This adds a generic tracing package to log messages to output if the target is set.

### CODE REVIEW

**Review by pjbgf — APPROVED** [review]
@aymanbagabas nice one, thanks for working on this. :bow:

**Comment by aymanbagabas** [comment]
Added test assertions and benchmarks

```
~/Source/go-git/go-git/utils/trace trace*
› go test -check.b -check.bmem -check.btime 5s
PASS: trace_test.go:63: TraceSuite.BenchmarkDisabledTarget      10000000000              1.96 ns/op       0 B/op           0 allocs/op
PASS: trace_test.go:69: TraceSuite.BenchmarkEnabledTarget       50000000               173 ns/op         14 B/op           1 allocs/op
OK: 2 passed
PASS
ok      github.com/go-git/go-git/v5/utils/trace 28.791s
```

File: utils/trace/trace_test.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+func (s *TraceSuite) BenchmarkDisabledTarget(c *C) {
+	for i := 0; i < c.N; i++ {
+		General.Print("test")
+	}
+}
+
+func (s *TraceSuite) BenchmarkEnabledTarget(c *C) {
+	SetTarget(General)
```
Comment (pjbgf) [inline utils/trace/trace_test.go:70]: For the purposes of the benchmark, I think the logger could write to `io.Discard` as we only want to understand the overall overhead.

```suggestion
	SetLogger(log.New(io.Discard, "", 0))
	SetTarget(General)
```

This does not seem to affect `check.v1` (as it probably wraps/ignore stderr), with standard benchmark, the call costs seems to be:
~1900 ns/op (default stderr)
~40 ns/op (io.Discard)
~300 ns/op (file on disk)

File: utils/trace/trace.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// Print prints the given message if tracing is enabled.
+func (t Target) Print(args ...interface{}) {
+	if int32(t)&current.Load() != 0 {
+		logger.Output(2, fmt.Sprint(args...)) // nolint: errcheck
+	}
+}
+
+// Printf prints the given message if tracing is enabled.
```
Comment (pjbgf) [inline utils/trace/trace.go:50]:
```suggestion
// Printf prints the given message only if the target is enabled.
```

File: utils/trace/trace.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+func SetTarget(target Target) {
+	current.Store(int32(target))
+}
+
+// SetLogger sets the logger to use for tracing.
+func SetLogger(l *log.Logger) {
+	logger = l
+}
+
+// Print prints the given message if tracing is enabled.
```
Comment (pjbgf) [inline utils/trace/trace.go:43]:
```suggestion
// Print prints the given message only if the target is enabled.
```

File: utils/trace/trace.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	// current is the targets that are enabled for tracing.
+	current atomic.Int32
+)
+
+func newLogger() *log.Logger {
+	return log.New(os.Stderr, "", log.Ltime|log.Lmicroseconds|log.Lshortfile)
+}
+
+// Target is a tracing target.
+type Target int
```
Comment (pjbgf) [inline utils/trace/trace.go:23]:
```suggestion
type Target int32
```

File: utils/trace/trace_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
+package trace
+
+import (
+	"bytes"
+	"log"
+	"testing"
+
+	. "gopkg.in/check.v1"
```
Comment (pjbgf) [inline utils/trace/trace_test.go:8]: For new code, we should move away from `check.v1` as that dependency is no longer maintained.

**Review by pjbgf — CHANGES_REQUESTED** [review]
Thank you for the benchmarks. Initially the costs for pktline seemed expensive, but upon looking into it more carefully it became clear that the `pktline` encoding is quite "cheap", therefore any additional call within that process would make a difference.

Below goes a few nits, otherwise LGTM

File: utils/trace/trace_test.go (hunk 5 above)
Comment (aymanbagabas) [reply utils/trace/trace_test.go]: switched to stdlib testing

**Comment by aymanbagabas** [comment]
> Thank you for the benchmarks. Initially the costs for pktline seemed expensive, but upon looking into it more carefully it became clear that the `pktline` encoding is quite "cheap", therefore any additional call within that process would make a difference.

Should we remove the printf calls then?

**Review by pjbgf — APPROVED** [review]
@aymanbagabas great job! Thanks for working on this. :trophy:

**Pull request merged**